'final' subclasses produce suboptimal self casts

Starting from:
```cython
# subclass.pyx
cdef class A:
    cdef int foo(self):
        return -1

    cdef int bar(self, int x):
        return -1

cdef class Sub(A):
    cdef int foo(self):
        return self.bar(0)

    cdef int bar(self, int x):
        return x + 1
```
compiles fine with `cythonize -b subclass.pyx`. If you add `@cython.final` onto `Sub`, then it produces a non-fatal warning:
```
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -m64 -fPIC -m64 -fPIC -fPIC -Iconda/envs/cp36/include/python3.6m -c subclass.c -o build/temp.linux-x86_64-3.6/subclass.o
subclass.c: In function ‘__pyx_f_8subclass_3Sub_foo’:
subclass.c:1667:41: warning: passing argument 1 of ‘__pyx_f_8subclass_3Sub_bar’ from incompatible pointer type [-Wincompatible-pointer-types]
   __pyx_r = __pyx_f_8subclass_3Sub_bar(((struct __pyx_obj_8subclass_A *)__pyx_v_self), 0);
                                        ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
subclass.c:1205:85: note: expected ‘struct __pyx_obj_8subclass_Sub *’ but argument is of type ‘struct __pyx_obj_8subclass_A *’
 static int __pyx_f_8subclass_3Sub_bar(CYTHON_UNUSED struct __pyx_obj_8subclass_Sub *__pyx_v_self, int __pyx_v_x); /* proto*/
                                                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~
```
But if you _also_ add `# distutils: language = c++` to the file to compile as C++, then it fails to compile:
```
gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -m64 -fPIC -m64 -fPIC -fPIC -Iconda/envs/cp36/include/python3.6m -c subclass.cpp -o build/temp.linux-x86_64-3.6/subclass.o
subclass.cpp: In function ‘int __pyx_f_8subclass_3Sub_foo(__pyx_obj_8subclass_Sub*)’:
subclass.cpp:1682:41: error: cannot convert ‘__pyx_obj_8subclass_A*’ to ‘__pyx_obj_8subclass_Sub*’
   __pyx_r = __pyx_f_8subclass_3Sub_bar(((struct __pyx_obj_8subclass_A *)__pyx_v_self), 0);
                                        ~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
subclass.cpp:1220:85: note:   initializing argument 1 of ‘int __pyx_f_8subclass_3Sub_bar(__pyx_obj_8subclass_Sub*, int)’
 static int __pyx_f_8subclass_3Sub_bar(CYTHON_UNUSED struct __pyx_obj_8subclass_Sub *__pyx_v_self, int __pyx_v_x); /* proto*/
                                                                                     ^
error: command 'gcc' failed with exit status 1
```

--- a/Cython/Compiler/ExprNodes.py
+++ b/Cython/Compiler/ExprNodes.py
@@ -6521,18 +6521,7 @@ class SimpleCallNode(CallNode):
                         goto_error = code.error_goto_if(" && ".join(exc_checks), self.pos)
                     else:
                         goto_error = ""
-                    undo_gh2747_hack = False
-                    if (self.function.is_attribute and
-                            self.function.entry and self.function.entry.final_func_cname and
-                            not code.funcstate.scope.is_cpp()
-                            ):
-                        # Hack around https://github.com/cython/cython/issues/2747
-                        # Disable GCC warnings/errors for wrong self casts.
-                        code.putln('#pragma GCC diagnostic ignored "-Wincompatible-pointer-types"')
-                        undo_gh2747_hack = True
                     code.putln("%s%s; %s" % (lhs, rhs, goto_error))
-                    if undo_gh2747_hack:
-                        code.putln("#pragma GCC diagnostic pop")
                 if self.type.is_pyobject and self.result():
                     self.generate_gotref(code)
             if self.has_optional_args:

--- a/Cython/Compiler/ModuleNode.py
+++ b/Cython/Compiler/ModuleNode.py
@@ -3935,7 +3935,8 @@ class ModuleNode(Nodes.Node, Nodes.BlockNode):
                 if entry.func_cname]
             if c_method_entries:
                 for meth_entry in c_method_entries:
-                    cast = meth_entry.type.signature_cast_string()
+                    vtable_type = meth_entry.vtable_type or meth_entry.type
+                    cast = vtable_type.signature_cast_string()
                     code.putln(
                         "%s.%s = %s%s;" % (
                             type.vtable_cname,

--- a/Cython/Compiler/Symtab.py
+++ b/Cython/Compiler/Symtab.py
@@ -242,6 +242,7 @@ class Entry(object):
     known_standard_library_import = None
     pytyping_modifiers = None
     enum_int_value = None
+    vtable_type = None
 
     def __init__(self, name, cname, type, pos = None, init = None):
         self.name = name
@@ -2551,6 +2552,9 @@ class CClassScope(ClassScope):
         if self.parent_type.is_final_type or entry.is_inline_cmethod or self.directives.get('final'):
             entry.is_final_cmethod = True
             entry.final_func_cname = entry.func_cname
+            if not type.is_fused:
+                entry.vtable_type = entry.type
+                entry.type = type
 
         return entry
 

--- a/tests/bugs.txt
+++ b/tests/bugs.txt
@@ -9,7 +9,6 @@ cfunc_call_tuple_args_T408
 generator_expressions_in_class
 for_from_pyvar_loop_T601
 temp_sideeffects_T654    # not really a bug, Cython warns about it
-inherited_final_method
 cimport_alias_subclass
 
 # PEP-489 is currently disabled

--- a/tests/run/inherited_final_method.pyx
+++ b/tests/run/inherited_final_method.pyx
@@ -3,32 +3,73 @@
 
 cimport cython
 
+cdef class TopBase:
+    cdef method(self):
+        return None
 
-cdef class BaseClass:
+
+cdef class BaseClass(TopBase):
     """
     >>> obj = BaseClass()
     >>> obj.call_base()
-    True
+    'base'
     """
+    cdef base(self):
+        return '@base'
+
     cdef method(self):
-        return True
+        return 'base'
 
     def call_base(self):
         return self.method()
 
 
 @cython.final
-cdef class Child(BaseClass):
+cdef class ChildClass(BaseClass):
     """
-    >>> obj = Child()
+    >>> obj = ChildClass()
     >>> obj.call_base()
-    True
+    'child'
     >>> obj.call_child()
-    True
+    'child'
     """
+    cdef child(self):
+        return '@child'
+
     cdef method(self):
-        return True
+        return 'child'
 
     def call_child(self):
         # original bug: this requires a proper cast for self
         return self.method()
+
+
+def test_BaseClass():
+    """
+    >>> test_BaseClass()
+    '@base'
+    'base'
+    """
+    cdef BaseClass obj = BaseClass()
+    print(repr(obj.base()))
+    print(repr(obj.method()))
+
+
+def test_ChildClass():
+    """
+    >>> test_ChildClass()
+    '@base'
+    '@child'
+    'child'
+    '@base'
+    '@child'
+    'child'
+    """
+    cdef ChildClass obj = ChildClass()
+    print(repr(obj.base()))
+    print(repr(obj.child()))
+    print(repr(obj.method()))
+    cdef BaseClass bobj = obj
+    print(repr(obj.base()))
+    print(repr(obj.child()))
+    print(repr(bobj.method()))

